jupyter_client/piper.py

```diff
@@ -0,0 +1,113 @@
+from codecs import getincrementaldecoder
+from itertools import groupby
+from operator import attrgetter
+
+from .session import new_id, msg_header
+
+BEGIN_LINE = b'\0JUPYTER STREAM BEGIN '
+END_LINE = b'\0JUPYTER STREAM END '
+BUFFER_LIMIT = 512
+
+class EndOfOutput(object):
+    """Marker for the end of output triggered by a parent message."""
+    def __init__(self, parent_id):
+        self.parent_id = parent_id
+
+class Output(object):
+    def __init__(self, text, parent_id):
+        self.text = text
+        self.parent_id = parent_id
+
+def concat_outputs(outputs):
+    for (cls, parent), group in groupby(outputs, key=lambda x: (type(x), x.parent_id)):
+        if cls is Output:
+            concat_text = ''.join(o.text for o in group)
+            yield Output(concat_text, parent)
+        else:
+            # There should never be more than one consecutive end marker
+            yield list(group)[0]
+
+class PipeCapturer(object):
+    """Translate output from a pipe into Jupyter stream messages."""
+    def __init__(self, stream_name='stdout', username='', session_id=''):
+        self.stream_name = stream_name
+        self.username = username
+        self.session_id = session_id
+        self.buffer = b''
+        self.decoder = getincrementaldecoder('utf-8')(errors='replace')
+        self.current_parent_id = None
+        self._queued_output = []
+
+    def new_stream_msg(self, text, parent_id):
+        msg = {}
+        header = msg_header(new_id(), 'stream', self.username, self.session_id)
+        parent_header = msg_header(parent_id, 'execute_request',
```

Comment:
I've lost track of what we actually wanted from the kernel nanny, so I thought I'd try to work on this bit separately, because at least I understand it. I thought we were heading in the direction of allowing a Python based process such as the notebook server integrate the nanny functionality without running a separate process, so I'd start by doing output capturing without the nanny inside one or more of our frontends. At the moment, I can't even think of an easy way to integrate it, though, because it requires changes in multiple repositories.